### PR TITLE
[WIP] Experimental FoW refactoring

### PR DESCRIPTION
**Pull request by ArseniyShestakov**
FoW code we have now is bad and solution I see for this is to have sight map for each team. So tiles within sight radius of owned objects would never ever covered by darkness and we don't need to re-check owned objects every time unless object themself changed.

How this should be implemented is another story. First I thought to store all tiles with sight in vector with duplicates and just add / remove them. Then I find out that current fogOfWarMap can be also used for storage since it's unlikely that we'll ever get to situation where ui8 limit wouldn't be enough.

### CODE REVIEW

File: lib/CGameState.cpp [hunk 1]
```diff
@@ -1608,6 +1610,64 @@ void CGameState::giveCampaignBonusToHero(CGHeroInstance * hero)
...
+	auto ts = getTeam(team);
+	std::unordered_set<int3, ShashInt3> tiles;
+	getTilesInRange(tiles, obj->getSightCenter(), obj->getSightRadius());
+	for(int3 t : tiles)
+	{
+		auto old = ts->fogOfWarMap[t.x][t.y][t.z];
+		if(add)
+		{
+			if(ts->fogOfWarMap[t.x][t.y][t.z] == 0)
+				ts->fogOfWarMap[t.x][t.y][t.z] = 2;
```
Comment (alexvins) [inline lib/CGameState.cpp]: Please explain why "0" become "2"

File: lib/CGameState.cpp [hunk 2]
```diff
@@ -1608,6 +1610,64 @@ void CGameState::giveCampaignBonusToHero(CGHeroInstance * hero)
...
+			else
+				ts->fogOfWarMap[t.x][t.y][t.z]++;
```
Comment (alexvins) [inline lib/CGameState.cpp]: Each element is ui8, you should check overflow somehow.

File: lib/CGameState.cpp (hunk 1 above)
Comment (ArseniyShestakov) [inline lib/CGameState.cpp]: 0 mean that tile in hidden from team.
1 mean that it's visible.
Anything more than 1 mean there is N objects with sight over that tile so it's can't be covered by darkness

File: lib/CGameState.cpp (hunk 2 above)
Comment (ArseniyShestakov) [inline lib/CGameState.cpp]: Yep, but possibility that we'll have more than 255 objects that have sight over one tile is very unlikely. Isn't it?

Comment (alexvins) [inline lib/CGameState.cpp]: Unlikely but possible

File: lib/CGameState.cpp (hunk 1 above)
Comment (alexvins) [inline lib/CGameState.cpp]: This should be documented. Also please add constants for 0, 1 & 2

File: lib/CGameState.cpp (hunk 2 above)
Comment (ArseniyShestakov) [inline lib/CGameState.cpp]: If it's happen that would mean FoW going to break completely.
So I can just add assert so it's indicate if something going really wrong.

**Comment by ArseniyShestakov** [comment]
So here is a bit cleaner version of code, but there still a lot of work to do:

* I'll rework map objects code related to sight range. At least where bonus system can be used.
* All situations where hero bonuses change must be handled properly too since both creatures and artifacts could increase sight radius using mods.

**Comment by ArseniyShestakov** [comment]
And now sadly I stuck a bit. I need proper way to detect if hero (and in future probably other object) bonuses changed so I can update sight map on the fly as well as sent some notification about that on client, but so far it's looks like there no easy way to do that. Of course addding code to every single netpack and GH function is not an option.

In past I had similar problem with movement-related bonuses since they can be affected by artifacts too.

File: lib/IGameCallback.cpp [hunk 3]
```diff
@@ -54,7 +54,7 @@ void CPrivilagedInfoCallback::getTilesInRange( std::unordered_set<int3, ShashInt
 		logGlobal->errorStream() << "Illegal call to getTilesInRange!";
 		return;
 	}
-	if (radious == -1) //reveal entire map
+	if (radious > 5000) //reveal entire map
```
Comment (alexvins) [inline lib/IGameCallback.cpp]: Should be constant.

**Comment by ArseniyShestakov** [comment]
So little update on this branch and why it's not going into next release. My idea with FoW map looks good and give more flexibility for the future multiplayer modes (like simultaneous turns), but it's have one important requirement: on change of any FoW-affecting bonuses gamestate must be update immidiately.

Otherwise there might be problem like that:

* Hero stopped on tile, FoW map altered in default 5 tiles around.  
* Player give hero sight radius altering artefact or hero got scouting skill.
* Then FoW map won't be updated of course.
* Now if hero is killed before he moved FoW map altered for more than 5 tiles around and corrupted.

We discussed this with Alex little bit before and end up we need proper events for bonus system so code like one that manages FoW can be automatically notified if any bonuses changed so it's can update player FoW map.

So for the next update I'll just add more plumbing for exist FoW code.

**Comment by alexvins** [comment]
Just ignore heroes when cashing FoW, apply hero sight radius OTF.

**Comment by ArseniyShestakov** [comment]
OTF? 

I can't "ignore" anything at all since fog of war map will be the only storage on FoW information and every other check will be removed.

**Comment by alexvins** [comment]
You should decide what is more difficult to implement: capture all bonus changes or do some OTF checks

**Comment by ArseniyShestakov** [comment]
In this case I want proper code with no compomises. After all old FoW code work mostly so no reason to replace one crutch other.

File: lib/mapObjects/CGTownInstance.cpp [hunk 4]
```diff
@@ -325,14 +325,7 @@ void CGDwelling::serializeJsonOptions(JsonSerializeFormat & handler)
...
 {
-	if (subID == ETownType::TOWER)
-	{
-		if (hasBuilt(BuildingID::GRAIL)) //skyship
-			return -1; //entire map
-		if (hasBuilt(BuildingID::LOOKOUT_TOWER)) //lookout tower
-			return 20;
-	}
-	return 5;
+	return 5 + valOfBonuses(Bonus::SIGHT_RADIOUS);
```
Comment (alexvins) [inline lib/mapObjects/CGTownInstance.cpp]: Add this "5" as permanent bonus

**Comment by alexvins** [comment]
Set of tiles visible to hero is actually a volatile variable. When you try to cache it you will get problems.

**Comment by ArseniyShestakov** [comment]
The problem is that I don't do any "caching" really. There no info who exactly see the tile in the map and all idea is just know that at least once object can "see" it so we never cover it by fog. Yeah it's possible to implement special hack to not deal with heroes or towns that way, but then there is no point in refactoring really since the idea is to simplify the FoW code.

**Comment by alexvins** [comment]
No, it is still a cache: does not matter that only count of uncovering object is stored. As I understand the problem, without cache we need to check all objects on map to see it tile can be hidden, reducing check to only heroes and may be towns is huge improvement anyway.

**Comment by ArseniyShestakov** [comment]
Ok I'll consider if it's possible to merge it with exception for heroes / towns.

**Comment by ArseniyShestakov** [comment]
So it's doesn't look that my idea would work after all. Problem is that our bonus system only make it possible to track when _something_ inside bonus tree has changed. So when it's changed I wouldn't be able to tell what bonus nodes are affected and even if it's was possible to track just single node it's possible that hero could for instance have artifact that affect not just his sight, but will also propagate it's bonus to the whole team or every owned object of player, etc.

So efficiently with every bonus tree change I'll need to rebuild FoW map one way or another.

**Pull request closed without merging**